### PR TITLE
Exercise 3

### PR DESCRIPTION
**Pull request by egrfly**
I've attempted some end-to-end testing using Cypress here which was quite good fun. These tests aren't currently using their own environment, so I had to write them in such a way as to avoid having any lasting impact on the main Trello board (by creating, updating and deleting an item all in one test). If I had more time, I would have ideally liked to create a system for spinning up boards specifically for the purpose of Cypress testing, but I decided not to prioritise this.

### CODE REVIEW

File: .gitignore
Comment (JackMead) [inline .gitignore:325]: No particular harm, but is there any reason you added in Node-related ignores?
Edit: Nevermind, found it!

File: README.md
Comment (JackMead) [inline README.md:111]: Thanks for the clear instructions, I've never actually used Cypress but this looks really cool!

One potential disadvantage over the suggested selenium approach is that this relies on the app already running and so misses out on the suggested automated test set up which includes creating a fresh throw-away Trello Board; so a bad test could easily end up leaving artefacts on the existing (hopefully at least dev/test!) board. With a more normal form of persistence we might solve that by spinning up the app against a throw-away database in a container or similar, but with Trello that might be more awkward. Not one to change this for, but worth bearing in mind!

File: cypress/e2e/index.cy.js
Comment (JackMead) [inline cypress/e2e/index.cy.js:47]: This seems really clean compared to what I'm used to with selenium!

File: todo_app/tests/integration/test_app_routes.py
Comment (JackMead) [inline todo_app/tests/integration/test_app_routes.py:1]: Minor, but personally I prefer tests to live at the top level of the project (often to mirror the source code structure when we have lots of unit tests) rather than under the todo_app folder. 

Separately, I'd advise adding an empty `__init__.py` file to each of these folders; Python is a little inconsistent about when it requires them but you may well find missing them bites you later when you move the tests inside Docker!

File: todo_app/tests/integration/test_app_routes.py
Comment (JackMead) [inline todo_app/tests/integration/test_app_routes.py:76]: Do these comments really imply that we want functions?

File: todo_app/tests/integration/test_app_routes.py
Comment (JackMead) [inline todo_app/tests/integration/test_app_routes.py:135]: These tests currently fail because (as I understand it!) your ItemIdConverter insists that item ids match a 24 length alphanumeric, but your test ids aren't of that form!

File: todo_app/tests/unit/test_items_view_model.py
Comment (JackMead) [inline todo_app/tests/unit/test_items_view_model.py:21]: Nice, super clean 👍

File: todo_app/templates/index.html
Comment (JackMead) [inline todo_app/templates/index.html:25]: This looks fantastic on the page!

File: todo_app/app.py
Comment (JackMead) [inline todo_app/app.py:43]: We could reduce this a bit by using the defaults I believe
```suggestion
            item.title = request.form.get('title', item.title)
            item.description = request.form.get('description', item.description)
            item.status = request.form.get('status', item.status)
```

**Review by JackMead — APPROVED** [review]
Nice work, this has some really great thorough tests - way beyond what we asked for! 😄 I've left a few optional suggestions below and in particular there's a minor error blocking two of your integration tests from passing, but once that's resolved I'm very happy for you to merge this when you're ready

File: README.md
Comment (egrfly) [reply README.md:111]: This was a drawback I was a bit unhappy with too. I decided not to prioritise sorting separate environments in the end, but it's something I would have liked to have done and may yet investigate more

File: todo_app/tests/integration/test_app_routes.py
Comment (egrfly) [reply todo_app/tests/integration/test_app_routes.py:1]: Thanks for the heads-up, will do that!

File: todo_app/tests/integration/test_app_routes.py
Comment (egrfly) [reply todo_app/tests/integration/test_app_routes.py:96]: Will do this too, seems like a good failsafe

File: todo_app/tests/integration/test_app_routes.py
Comment (egrfly) [reply todo_app/tests/integration/test_app_routes.py:135]: Whoops, forgot to rerun the tests after merging main! Not ideal

File: todo_app/templates/index.html
Comment (egrfly) [reply todo_app/templates/index.html:25]: Thanks!

File: todo_app/app.py
Comment (egrfly) [reply todo_app/app.py:43]: Thanks, that's much nicer. I was originally using the walrus operator but decided to remove it to keep the Python 3.7 compatibility. This is exactly the kind of thing I was after

File: todo_app/tests/integration/test_app_routes.py
Comment (egrfly) [reply todo_app/tests/integration/test_app_routes.py]: Really sorry, not entirely sure what you mean here!

**Pull request merged**
